plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui.test/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/internal/DeployArtifactValidatorTest.java

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.appengine.deploy.ui.internal;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.io.IOException;
+import org.eclipse.core.databinding.observable.value.IObservableValue;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Path;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DeployArtifactValidatorTest {
+
+  @Rule public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  @Mock private IObservableValue deployArtifactPath;
+
+  private IPath basePath;
+  private DeployArtifactValidator pathValidator;
+
+  @Before
+  public void setUp() throws IOException {
+    basePath = new Path(tempFolder.newFolder().toString());
+    when(deployArtifactPath.getValueType()).thenReturn(String.class);
+    pathValidator = new DeployArtifactValidator(basePath, deployArtifactPath);
+    assertTrue(basePath.isAbsolute());
+  }
+
+  @Test
+  public void testConstructor_nonAbsoluteBasePath() {
+    try {
+      when(deployArtifactPath.getValue()).thenReturn("some.jar");
+      new DeployArtifactValidator(new Path("non/absolute/base/path"), deployArtifactPath);
+      fail();
+    } catch (IllegalArgumentException ex) {
+      assertEquals("basePath is not absolute.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testValidate_emptyField() {
+    when(deployArtifactPath.getValue()).thenReturn("");
+
+    IStatus result = pathValidator.validate();
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("Missing WAR or JAR path.", result.getMessage());
+  }
+
+  @Test
+  public void testValidate_relativePathAndNoArtifact() {
+    when(deployArtifactPath.getValue()).thenReturn("relative/path/some.jar");
+
+    IStatus result = pathValidator.validate();
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("WAR or JAR does not exist.", result.getMessage());
+  }
+
+  @Test
+  public void testValidate_absolutePathAndNoArtifact() {
+    String absolutePath = basePath + "/sub/directory/some.jar";
+    when(deployArtifactPath.getValue()).thenReturn(absolutePath);
+
+    IStatus result = pathValidator.validate();
+    assertEquals(IStatus.ERROR, result.getSeverity());
+    assertEquals("WAR or JAR does not exist.", result.getMessage());
```

Comment:
This message could be "JAR does not exist" since we know it's a jar that's be looked for. More specific error messages are more helpful to the user.

Reply:
The global deploy dialog (hence this validator) will accept both WAR and JAR, so this will require an additional branch to determine if it's a WAR or a JAR. I think I should it to "File does not exist: {0}" (showing the actual path entered together). This message is a remnant of the app.yaml counterpart, where the file name has to be always `app.yaml`.